Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp

```diff
@@ -123,8 +123,8 @@ bool LegacyRenderSVGEllipse::shapeDependentStrokeContains(const FloatPoint& poin
 {
     // The optimized code below does not support non-smooth strokes so we need to
     // fall back to LegacyRenderSVGShape::shapeDependentStrokeContains in these cases.
-    if (!hasSmoothStroke() && !hasPath())
-        LegacyRenderSVGShape::updateShapeFromElement();
+    if (!hasSmoothStroke())
+        ensurePath();
```

Comment:
Who is going to calculate the `m_fillBoundingBox` and `m_strokeBoundingBox`? I think is a weird state of the renderer. The path is created from the element but the bounding boxes are empty.

Reply:
`m_fillBoundingBox` and `m_strokeBoundingBox` computation is the responsibility of `updateShapeFromElement`, not in this function. So when calling `shapeDependentStrokeContains`, it already has these information computed from `updateShapeFromElement`, this is an invariant of this class.

In this function, we are using `m_center` in `!hasPath()` case, and this `m_center` is computed  in `updateShapeFromElement` (which is calling `calculateRadiiAndCenter` to compute it). This is because `updateShapeFromElement ` is the function which updates these values (including `m_strokeBoundingBox` etc.).

Another example. In `LegacyRenderSVGShape::strokeContains`, it is already touching `m_strokeBoundingBox` via `strokeBoundingBox()`. This is because this function is relying on `updateShapeFromElement`, which offers `m_fillBoundingBox` and `m_strokeBoundingBox`.

We are not always computing these values from the path since some of shape can compute them without having a path. `LegacyRenderSVGEllipse::updateShapeFromElement` is one of the example, which is computing `m_fillBoundingBox` and `m_strokeBoundingBox` without creating a path.

The reason why we need to have a path here is that we need path based approach for non-smooth-stroke and use `LegacyRenderSVGShape::shapeDependentStrokeContains`. Our current code is intentionally not creating `m_path` for non-smooth-stroke in `LegacyRenderSVGEllipse::updateShapeFromElement` because the goal of that function is computing `m_fillBoundingBox` and `m_strokeBoundingBox`, and that does not require a path for non-smooth-stroke (so our current code has optimization for that). And only this case needs a path, but bounding boxes are already computed correctly by `LegacyRenderSVGEllipse::updateShapeFromElement` (And it is not `shapeDependentStrokeContains`'s responsibility).

So, this patch makes it more consistent behavior. Here we only materialize a path because it is necessary for stroke-based contains operations in `LegacyRenderSVGShape::shapeDependentStrokeContains`. But we do not need `m_fillBoundingBox` and `m_strokeBoundingBox`. They are computed in `LegacyRenderSVGEllipse::updateShapeFromElement`.

Reply:
After this patch there will be no explicit calls to `LegacyRenderSVGShape::updateShapeFromElement` from functions other than `updateShapeFromElement` overrides, is that right? I think that's a positive change since we'll no longer be using it for two purposes (i.e. the main purpose of updating shape data from layout, and this secondary purpose of ensuring the Path exists in functions like this one that happen to need it).

Reply:
Right! That's the main purpose of this patch. After this patch, we no longer call `LegacyRenderSVGShape::updateShapeFromElement` from non `updateShapeFromElement` places. That makes these responsibility, updating m_strokeBoundingBox and m_fillBoundingBox, explicit and clear.

This patch paves a way to lazily compute `m_strokeBoundingBox` by making `updateShapeFromElement` more and more light weight (this is the next patch), and the final patch will start avoiding `m_strokeBoundingBox` until it becomes really necessary.

Also, that change will effectively make `LegacyRenderSVGShape::updateShapeFromElement` just `m_fillBoundingBox = ensurePath().boundingBox();`. Probably I'll make `LegacyRenderSVGShape::updateShapeFromElement` as pure virtual function, and moving them to each derived `updateShapeFromElement` function (because it now just becomes `m_fillBoundingBox = ensurePath().boundingBox();`, then each derived shape's `updateShapeFromElement` becomes really concise, easy to read, and clear.